File: src/psk/pskwebsite/psk/server/src/assistant/actions.js
/**
 * actions.js — turns an entity_id into a typed, executable action.
 *
 * SAFETY INVARIANT: this module is the ONLY place a route is produced, and it
 * produces routes exclusively by looking up the entity registry. There is no
 * function here that accepts a URL string from a model or from user input.
 * That is what makes an invented link structurally impossible rather than
 * merely discouraged by a prompt.
 */
import { getEntity } from './resolver.js';
import config from './config.js';

/** intent + entity type -> action name */
const ACTION_FOR_TYPE = {
  GAME: 'OPEN_GAME',
  CATEGORY: 'OPEN_CATEGORY',
  SPORT: 'OPEN_SPORT',
  PAGE: 'OPEN_PAGE',
  ACCOUNT: 'OPEN_PAGE',
  POLICY: 'OPEN_PAGE',
  SUPPORT: 'OPEN_PAGE',
};

const ACTION_FOR_INTENT = {
  DEPOSIT: 'OPEN_DEPOSIT',
  WITHDRAW: 'OPEN_WITHDRAW',
  BET_HISTORY: 'OPEN_BET_HISTORY',
  KYC: 'OPEN_KYC',
  PROMOTION: 'OPEN_PROMOTIONS',
  SUPPORT: 'OPEN_SUPPORT',
  RESPONSIBLE_GAMING: 'OPEN_RESPONSIBLE_GAMING',
};

const LABEL_VERB = {
  OPEN_GAME: 'Play',
  OPEN_CATEGORY: 'Browse',
  OPEN_SPORT: 'Open',
  OPEN_DEPOSIT: 'Go to',
  OPEN_WITHDRAW: 'Go to',
  OPEN_BET_HISTORY: 'View',
  OPEN_KYC: 'Open',
  OPEN_PROMOTIONS: 'View',
  OPEN_SUPPORT: 'Contact',
  OPEN_RESPONSIBLE_GAMING: 'Open',
  OPEN_PAGE: 'Open',
};

/**
 * Build an action from an entity_id. Returns null if the id is unknown —
 * an unknown id must never produce a guessed route.
 */
export function buildAction(entityId, { intent = null, label = null } = {}) {
  const e = getEntity(entityId);
  if (!e) return null;

  const action = ACTION_FOR_INTENT[intent] || ACTION_FOR_TYPE[e.type] || 'OPEN_PAGE';
  const verb = LABEL_VERB[action] || 'Open';

  return {
    action,
    label: label || `${verb} ${e.name}`,
    entity_id: e.id,
    entity_type: e.type,
    route: e.route,                       // from the registry, never generated
    requires_auth: Boolean(e.requires_auth),
    requires_kyc: Boolean(e.requires_kyc),
    requires_confirmation: config.confirmActions.has(action),
    status: e.status,
  };
}

/** Several actions at once, skipping any unknown ids. */
export function buildActions(entityIds, opts = {}) {
  return entityIds.map((id) => buildAction(id, opts)).filter(Boolean);
}

/**
 * Decide whether an action may execute now.
 * Returns { allowed, reason, gate } — the caller turns a gate into a prompt.
 */
export function checkGates(action, user = {}) {
  if (!action) return { allowed: false, reason: 'unknown_entity' };

  if (action.status && action.status !== 'active') {
    return { allowed: false, reason: 'entity_inactive', gate: 'unavailable' };
  }
  if (action.requires_auth && !user.authenticated) {
    return { allowed: false, reason: 'requires_login', gate: 'login' };
  }
  if (action.requires_kyc && user.kyc_status !== 'verified') {
    return { allowed: false, reason: 'requires_kyc', gate: 'kyc' };
  }
  return { allowed: true };
}

/** Contextual follow-up actions worth surfacing under an answer. */
export function supportingActions(intent, primaryId) {
  const ids = [];
  if (intent === 'RESPONSIBLE_GAMING') {
    ids.push('page_responsible_gaming', 'page_contact');
  } else if (intent === 'WITHDRAW') {
    // this build has no withdrawal or KYC flow, so point at account + support
    ids.push('page_account', 'page_contact');
  } else if (intent === 'DEPOSIT') {
    ids.push('page_responsible_gaming');
  } else if (intent === 'GAME_INFO') {
    ids.push('page_responsible_gaming');
  } else if (intent === 'ACCOUNT_DATA') {
    ids.push('page_account', 'page_tickets');
  } else if (intent === 'FAQ' || intent === 'SUPPORT') {
    ids.push('page_help', 'page_contact');
  }
  return buildActions(ids.filter((id) => id !== primaryId));
}

export default { buildAction, buildActions, checkGates, supportingActions };

File: src/psk/pskapp/lib/assistant/actions.dart
import 'assistant_models.dart';
import 'resolver.dart';

/// Turns an entity_id into a typed, executable action.
///
/// SAFETY INVARIANT: this is the ONLY place a route is produced, and it does so
/// exclusively by looking up the entity registry. Nothing here accepts a route
/// string from a model or from user input.
class ActionBuilder {
  static const Map<String, String> _actionForType = {
    'GAME': 'OPEN_GAME',
    'CATEGORY': 'OPEN_CATEGORY',
    'SPORT': 'OPEN_SPORT',
    'PAGE': 'OPEN_PAGE',
  };

  static const Map<String, String> _actionForIntent = {
    'DEPOSIT': 'OPEN_DEPOSIT',
    'WITHDRAW': 'OPEN_WITHDRAW',
    'BET_HISTORY': 'OPEN_BET_HISTORY',
    'KYC': 'OPEN_KYC',
    'PROMOTION': 'OPEN_PROMOTIONS',
    'SUPPORT': 'OPEN_SUPPORT',
    'RESPONSIBLE_GAMING': 'OPEN_RESPONSIBLE_GAMING',
  };

  static const Map<String, String> _labelVerb = {
    'OPEN_GAME': 'Play',
    'OPEN_CATEGORY': 'Browse',
    'OPEN_SPORT': 'Open',
    'OPEN_DEPOSIT': 'Go to',
    'OPEN_WITHDRAW': 'Go to',
    'OPEN_BET_HISTORY': 'View',
    'OPEN_KYC': 'Open',
    'OPEN_PROMOTIONS': 'View',
    'OPEN_SUPPORT': 'Contact',
    'OPEN_RESPONSIBLE_GAMING': 'Open',
    'OPEN_PAGE': 'Open',
  };

  /// Financial destinations always confirm, whatever the confidence.
  static const Set<String> confirmActions = {'OPEN_DEPOSIT', 'OPEN_WITHDRAW'};

  final EntityResolver resolver;
  ActionBuilder(this.resolver);

  AssistantAction? build(String entityId, {String? intent, String? label}) {
    final e = resolver.getEntity(entityId);
    if (e == null) return null;
    final action = (intent != null ? _actionForIntent[intent] : null) ?? _actionForType[e.type] ?? 'OPEN_PAGE';
    final verb = _labelVerb[action] ?? 'Open';
    return AssistantAction(
      action: action,
      label: label ?? '$verb ${e.name}',
      entityId: e.id,
      entityType: e.type,
      route: e.route,
      requiresAuth: e.requiresAuth,
      requiresConfirmation: confirmActions.contains(action),
      active: e.active,
    );
  }

  List<AssistantAction> buildAll(Iterable<String> ids, {String? intent}) =>
      ids.map((id) => build(id, intent: intent)).whereType<AssistantAction>().toList();

  /// Contextual follow-up actions worth surfacing under an answer.
  List<AssistantAction> supporting(String intent, String? primaryId) {
    final ids = <String>[];
    switch (intent) {
      case 'RESPONSIBLE_GAMING':
        ids.addAll(['page_responsible_gaming', 'page_contact']);
      case 'WITHDRAW':
        ids.addAll(['page_account', 'page_contact']);
      case 'DEPOSIT':
      case 'GAME_INFO':
        ids.add('page_responsible_gaming');
      case 'ACCOUNT_DATA':
        ids.addAll(['page_account', 'page_tickets']);
      case 'FAQ':
      case 'SUPPORT':
        ids.addAll(['page_help', 'page_contact']);
    }
    return buildAll(ids.where((id) => id != primaryId));
  }
}

class GateResult {
  final bool allowed;
  final String? gate; // login | unavailable
  const GateResult(this.allowed, [this.gate]);
}

GateResult checkGates(AssistantAction? action, {required bool authenticated}) {
  if (action == null) return const GateResult(false, 'unknown');
  if (!action.active) return const GateResult(false, 'unavailable');
  if (action.requiresAuth && !authenticated) return const GateResult(false, 'login');
  return const GateResult(true);
}
